### PR TITLE
Permit safe unsigned and signed conversions

### LINKED ISSUE #54: Overly restrictive conversion between integer instantiations
Hi Jonathan!

Looking at your documentation for *type_safe* and saw this about [`integer`](http://foonathan.net/doc/type_safe/doc_integer.html#ts::integer-IntegerT,Policy-):

> A conversion is considered safe if both integer types have the same signedness and the size of the value being converted is less than or equal to the destination size.

But what if you are converting, say, from `uint16_t` to `int32_t`?  Is this not allowed?  Is it not safe?

BTW, I have a [similar test](https://github.com/johnmcfarlane/cnl/blob/305801a35a5ca01e9983e93dec270f83321d280e/include/cnl/num_traits.h#L272) with no such restriction.

Cheers,
John

### PR DESCRIPTION
**Pull request by Xenoamor**
This is to resolve #54

### CODE REVIEW

**Comment by foonathan** [comment]
Could you add a test for your truth table with `std::is_convertible`? Because I don't think your condition is correct, but could be wrong.

**Comment by Xenoamor** [comment]
Unit tests are always worth doing. Cheers for pointing the lack of it out
Can't say I'm a fan of these GCC warnings. Should I disable them in the unit test?

File: test/integer.cpp [hunk 1]
```diff
@@ -295,4 +295,108 @@ TEST_CASE("integer")
         ia = type_safe::abs(i);
         REQUIRE(static_cast<unsigned>(ia) == 123u);
     }
+    SECTION("signed_to_unsigned")
+    {
+        // From int8_t
+        REQUIRE( !std::is_convertible<integer<int8_t>, integer<uint8_t>>::value );
+        REQUIRE(  std::is_convertible<integer<int8_t>, integer<uint16_t>>::value );
```
Comment (foonathan) [inline test/integer.cpp]: This test says that we can convert `int8_t` to `uint16_t`, which is wrong because of negative numbers.

Comment (foonathan) [reply test/integer.cpp]: This is also the cause of the warning.

**Comment by Xenoamor** [comment]
Interesting.. any idea why MSVC might be failing here?

**Comment by Xenoamor** [comment]
[Could be this?](https://devblogs.microsoft.com/cppblog/visual-c-2015-update-2-bug-fixes/)
image

**Comment by foonathan** [comment]
Debugging on godbolt lead me to issues with deleted constructors, so I've disabled them under old MSVC. They're merely there to improve the error messages anyway.

**Comment by foonathan** [comment]
Anyways, thanks for your help!

**Pull request merged**